mitxgraders/baseclasses.py

```diff
@@ -205,6 +207,60 @@ def log_output(self):
         content = "\n".join(self.debuglog)
         return "<pre>{content}</pre>".format(content=content)
 
+    @classmethod
+    def ensure_text_inputs(cls, student_input, allow_lists=True, allow_single=True):
+        """
+        Ensures that student_input is a list of text strings.
+
+        NOTE: Overriden by ItemGrader to ensure input is a single text string.
+        """
+        list_errmsg = None
+        single_errmsg = None
+
+        if allow_lists and isinstance(student_input, list):
+            try:
+                return Schema([text_string])(student_input)
+            except MultipleInvalid as error:
+                pos = error.path[0]
+                list_errmsg = (
+                    "Expected a list of text strings for student_input, but "
+                    "item at position {pos} has {thetype}"
+                    .format(pos=pos, thetype=type(student_input[pos])))
+
+        if allow_lists and not isinstance(student_input, list):
+            list_errmsg = (
+                "Expected student_input to be a list of text strings, but "
+                "received {}".format(type(student_input)))
+
+        if allow_single:
+            try:
+                return Schema(text_string)(student_input)
+            except MultipleInvalid as error:
+                single_errmsg = (
```

Comment:
Again, raise immediately.

Reply:
Or is the point not to raise this if both `allow_lists` and `allow_single` are True? Still, seems like too much spaghetti...

Perhaps something like the following?
```
        if allow_lists and isinstance(student_input, list):
            try:
                return Schema([text_string])(student_input)
            except MultipleInvalid as error:
                pass  # Will raise error below
        if allow_single:
            try:
                return Schema(text_string)(student_input)
            except MultipleInvalid as error:
                pass  # Will raise error below
```

This parses the valid inputs immediately, and you can select the appropriate error message to raise afterwards.

Reply:
> Or is the point not to raise this if both allow_lists and allow_single are True? Still, seems like too much spaghetti...

Yes, that's why I avoided raising immediately.

I'm not sure that the structure you're thinking of will work without nesting the try/excepts fairly heavily, since the error variables are local:

```python
try:
    1/0
except ZeroDivisionError as error1:
    print(error1) # prints division by zero
    pass

try:
    import puppy
except ImportError as error2:
    print(error2) # pritns No module named 'puppy'
    pass

print(error1) # raises NameError: name 'error1' is not defined
print(error2)
```

Let me see if I can think about other ways to restructure this.

Reply:
Well, I think the new version is cleaner, but I'm sure you'll like it more. If you have other ideas, or if I misunderstood your bit about saving and raising errors, let me know.